### PR TITLE
IGAPP-837: Update typescript to 4.4

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
     "noUncheckedIndexedAccess": true,
     "resolveJsonModule": true,
     "strict": true,
-    "target": "ESNext"
+    "target": "ESNext",
+    "useDefineForClassFields": false
```
Comment (klinzo) [inline tsconfig.json:13]: Why is this change necessary?

File: web/src/utils/retryImport.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ const retry = async <T>(
 ): Promise<{ default: ComponentType<T> }> => {
   try {
     return componentImport()
-  } catch (error) {
+  } catch (error: any) {
```
Comment (klinzo) [inline web/src/utils/retryImport.ts:22]: I'd favor using `unknown` as type annotation, `any` is generally not a recommended type annotation. Same for other usages. We also have a linting rule for this (which is only a warning right now). Maybe we should change that.

File: web/src/utils/retryImport.ts [hunk 3]
```diff
@@ -19,7 +19,7 @@ const retry = async <T>(
...
     await wait(interval)
     if (retriesLeft === 0) {
       throw new Error(error)
```
Comment (klinzo) [inline web/src/utils/retryImport.ts:25]:
```suggestion
      throw error
```

File: native/src/utils/sendTrackingSignal.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ export const sendRequest = async (signal: SignalType): Promise<void> => {
       await createTrackingEndpoint().request(signal)
     }
   } catch (e) {
-    if (fromError(e) === ErrorCode.NetworkConnectionFailed) {
+    if (e instanceof Error && fromError(e) === ErrorCode.NetworkConnectionFailed) {
```
Comment (klinzo) [inline native/src/utils/sendTrackingSignal.ts:21]: I think with your change to `fromError` this additional check should not be needed, right?

File: api-client/src/errors/__tests__/NotFoundError.spec.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ describe('NotFoundError', () => {
         city: 'augsburg',
         language: 'language'
       })
-    } catch (error) {
+    } catch (error: any) {
```
Comment (klinzo) [inline api-client/src/errors/__tests__/NotFoundError.spec.ts:23]: Please use `unknown` instead.

File: api-client/src/endpoints/hooks/useLoadFromEndpoint.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ export const loadFromEndpoint = async <T>(
       setError(null)
     }
   } catch (e) {
-    setError(e)
+    setError(e instanceof Error ? e : null)
```
Comment (klinzo) [inline api-client/src/endpoints/hooks/useLoadFromEndpoint.ts:26]: I don't think we should ignore errors if it is not an instance of type Error. Maybe just use
```suggestion
    setError(e instanceof Error ? e : new Error())
```

File: tsconfig.json (hunk 1 above)
Comment (SteffiStuffel) [reply tsconfig.json:13]: `In 2021 the class fields feature was added into the JavaScript specification with behavior which differed from how TypeScript had implemented it. In preparation for this, in TypeScript 3.7, a flag was added (useDefineForClassFields) to migrate to emitted JavaScript to match the JavaScript standard behavior.

Now that the feature is in JavaScript we are changing the default to true for ES2022 and above, including ESNext.`

See more https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier

File: web/src/utils/retryImport.ts (hunk 2 above)
Comment (SteffiStuffel) [reply web/src/utils/retryImport.ts:22]: I had to add `any` explicitly because typescript changed it standard behavior from `any` to `unknown` leading to errors, as properties are not available on unknown, etc.

**Review by klinzo — APPROVED** [review]
not tested

**Review by f1sh1918 — APPROVED** [review]
